x-pack/plugin/deprecation/src/main/java/org/elasticsearch/xpack/deprecation/MlDeprecationChecker.java

```diff
@@ -0,0 +1,79 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.deprecation;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.XPackSettings;
+import org.elasticsearch.xpack.core.deprecation.DeprecationIssue;
+import org.elasticsearch.xpack.core.ml.action.GetDatafeedsAction;
+import org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfig;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+public class MlDeprecationChecker implements DeprecationChecker {
+
+    static Optional<DeprecationIssue> checkDataFeedQuery(DatafeedConfig datafeedConfig, NamedXContentRegistry xContentRegistry) {
+        List<String> deprecations = datafeedConfig.getQueryDeprecations(xContentRegistry);
+        if (deprecations.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(new DeprecationIssue(DeprecationIssue.Level.WARNING,
+                "Datafeed [" + datafeedConfig.getId() + "] uses deprecated query options",
+                "https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html#breaking_70_search_changes",
+                deprecations.toString()));
+        }
+    }
+
+    static Optional<DeprecationIssue> checkDataFeedAggregations(DatafeedConfig datafeedConfig, NamedXContentRegistry xContentRegistry) {
+        List<String> deprecations = datafeedConfig.getAggDeprecations(xContentRegistry);
+        if (deprecations.isEmpty()) {
+            return Optional.empty();
+        } else {
+            return Optional.of(new DeprecationIssue(DeprecationIssue.Level.WARNING,
+                "Datafeed [" + datafeedConfig.getId() + "] uses deprecated aggregation options",
+                "https://www.elastic.co/guide/en/elasticsearch/reference/master/breaking-changes-7.0.html" +
+                    "#breaking_70_aggregations_changes", deprecations.toString()));
+        }
+    }
+
+    @Override
+    public boolean enabled(Settings settings) {
+        return XPackSettings.MACHINE_LEARNING_ENABLED.get(settings);
+    }
+
+    @Override
+    public void check(DeprecationCheckerComponents components, ActionListener<CheckResult> deprecationIssueListener) {
+        if (XPackSettings.MACHINE_LEARNING_ENABLED.get(components.settings()) == false) {
+            deprecationIssueListener.onResponse(new CheckResult(getName(), Collections.emptyList()));
+        }
+        Client client = components.client();
+        ClientHelper.executeAsyncWithOrigin(client, ClientHelper.DEPRECATION_ORIGIN, GetDatafeedsAction.INSTANCE,
```

Comment:
Is there a good reason not to construct the `Components` with an `OriginSettingClient`, so each plugin doesn't have to do this wrapping itself?